inst/include/dust/interface.hpp

```diff
@@ -406,6 +410,56 @@ cpp11::sexp dust_info(const typename T::init_t& pars) {
   return R_NilValue;
 }
 
+template <typename T, typename std::enable_if<!std::is_same<dust::no_data, typename T::data_t>::value, int>::type = 0>
+void dust_set_data(SEXP ptr, cpp11::list r_data) {
+  typedef typename T::data_t data_t;
+  Dust<T> *obj = cpp11::as_cpp<cpp11::external_pointer<Dust<T>>>(ptr).get();
+  std::map<size_t, data_t> data;
+  size_t len = r_data.size();
+  for (size_t i = 0; i < len; ++i) {
+    cpp11::list el = r_data[i];
+    const size_t step_i = cpp11::as_cpp<int>(el[0]);
+    const data_t data_i = dust_data<T>(cpp11::as_cpp<cpp11::list>(el[1]));
+    data[step_i] = data_i;
+  }
+  obj->set_data(data);
+}
+
+template <typename T, typename std::enable_if<!std::is_same<dust::no_data, typename T::data_t>::value, int>::type = 0>
+cpp11::sexp dust_compare_data(SEXP ptr) {
+  Dust<T> *obj = cpp11::as_cpp<cpp11::external_pointer<Dust<T>>>(ptr).get();
+  std::vector<double> ret = obj->compare_data();
+  if (ret.size() == 0) {
+    return R_NilValue;
+  }
+  cpp11::writable::doubles ret_r(ret.size());
+  std::copy(ret.begin(), ret.end(), REAL(ret_r));
+  return ret_r;
+}
+
+template <typename T, typename std::enable_if<std::is_same<dust::no_data, typename T::data_t>::value, int>::type = 0>
+void dust_set_data(SEXP ptr, cpp11::list r_data) {
+  cpp11::stop("The 'set_data' method is not supported for this class");
+}
+
+template <typename T, typename std::enable_if<std::is_same<dust::no_data, typename T::data_t>::value, int>::type = 0>
+cpp11::sexp dust_compare_data(SEXP ptr) {
+  cpp11::stop("The 'compare_data' method is not supported for this class");
+  return R_NilValue; // never gets here
+}
```

Comment:
Perhaps just make a comment above these classes to note that the template metaprogramming leads to compilation of this function if data is null/missing; the full function above if instead it's given